Source/WebCore/css/StyleProperties.cpp

```diff
@@ -581,180 +581,211 @@ String StyleProperties::fontSynthesisValue() const
 }
 
 String StyleProperties::getLayeredShorthandValue(const StylePropertyShorthand& shorthand) const
 {
-    StringBuilder result;
-
-    const unsigned size = shorthand.length();
-    Vector<RefPtr<CSSValue>> values(size);
-    size_t numLayers = 0;
+    ASSERT(shorthand.length() <= 9);
 
-    for (unsigned i = 0; i < size; ++i) {
-        values[i] = getPropertyCSSValue(shorthand.properties()[i]);
-        if (values[i]->isBaseValueList())
-            numLayers = std::max(downcast<CSSValueList>(*values[i]).length(), numLayers);
-        else
-            numLayers = std::max<size_t>(1U, numLayers);
+    unsigned length = shorthand.length();
+    Vector<Ref<CSSValue>> values;
+    values.reserveInitialCapacity(length);
+    size_t numLayers = 1;
+    for (unsigned i = 0; i < length; ++i) {
+        auto value = getPropertyCSSValue(shorthand.properties()[i]);
+        if (value->isBaseValueList())
+            numLayers = std::max(downcast<CSSValueList>(*value).length(), numLayers);
+        values.uncheckedAppend(value.releaseNonNull());
     }
 
-    // Now stitch the properties together.
-    // Implicit initial values are flagged as such and can safely be omitted.
+    StringBuilder result;
     for (size_t i = 0; i < numLayers; i++) {
-        StringBuilder layerResult;
-        bool useRepeatXShorthand = false;
-        bool useRepeatYShorthand = false;
-        bool useSingleWordShorthand = false;
-        bool foundPositionYCSSProperty = false;
-        for (unsigned j = 0; j < size; j++) {
-            auto property = shorthand.properties()[j];
-
-            auto value = values[j];
-            if (value) {
-                if (value->isBaseValueList())
-                    value = downcast<CSSValueList>(*value).item(i);
-                else {
-                    // Color only belongs in the last layer.
-                    if (property == CSSPropertyBackgroundColor) {
-                        if (i != numLayers - 1)
-                            value = nullptr;
-                    } else if (i) // Other singletons only belong in the first layer.
-                        value = nullptr;
+        RefPtr<CSSValue> layerValues[9];
+        bool skipSerializingLayerValues[9] { };
+
+        ASSERT(length <= std::size(layerValues));
+
+        for (unsigned j = 0; j < length; j++) {
+            auto longhand = shorthand.properties()[j];
+            auto& value = values[j].get();
+            RefPtr<CSSValue> layerValue;
+            bool skipSerializing = false;
+            if (value.isBaseValueList())
+                layerValue = downcast<CSSValueList>(value).item(i);
+            else {
+                layerValue = &value;
+                if (longhand != CSSPropertyBackgroundColor) {
+                    // Singletons are only in the first layer.
+                    if (i)
+                        skipSerializing = true;
+                } else {
+                    // Color is only in the last layer.
```

Comment:
Anyway to statically assert this? And if not currently, is there some metadata we should consider exposing from the generators to make this statically assertable?

Reply:
This is not new. I wrote the code in an easier to read way, but the logic was there before. I’m not sure there’s anything to assert really. I’m not going to change anything here.

Reply:
Maybe you can figure out a way to improve this later, but I was really just trying to preserve what was there already in a style that was easy enough to read.

Reply:
This rule about what is in what layer comes from the specification text and is not present in the BNF as far as I can tell.